File: finalyearproject/repos/dkd-platform-libs/sdk/go/custody_hash.go
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// CustodyHash Specification v2 — DM §2 (RFC-8785 subset R1-R9). One of five byte-identical
// runtime implementations; the shared gate is sdk/testvectors/custodyhash_vectors.json (PL-01).
// Ported verbatim from the proven custody-ledger-svc/internal/custody/canonical.go reference.
package dkdplatform

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"math"
	"sort"
	"strconv"
)

// CustodyCanonicalJSON serializes v per CustodyHash Spec v2 rules R1-R9:
// R2 null/absent omission (object members only), R3 lexicographic UTF-8 key order at every
// depth, R4 no whitespace, R5 UTF-8 strings without \uXXXX for code points < U+0080,
// R6 integers as plain decimal, R7 lowercase booleans, R8 arrays in declaration order,
// R9 recursive application.
func CustodyCanonicalJSON(v any) ([]byte, error) {
	var buf bytes.Buffer
	if err := writeCustodyCanonical(&buf, v); err != nil {
		return nil, err
	}
	return buf.Bytes(), nil
}

func writeCustodyCanonical(buf *bytes.Buffer, v any) error {
	switch x := v.(type) {
	case nil:
		return fmt.Errorf("custody: canonical JSON forbids null values outside omitted object members (R2)")
	case map[string]any:
		keys := make([]string, 0, len(x))
		for k, val := range x {
			if val == nil {
				continue // R2: null members are omitted
			}
			keys = append(keys, k)
		}
		sort.Strings(keys) // R3: lexicographic byte order
		buf.WriteByte('{')
		for i, k := range keys {
			if i > 0 {
				buf.WriteByte(',')
			}
			if err := writeCustodyString(buf, k); err != nil {
				return err
			}
			buf.WriteByte(':')
			if err := writeCustodyCanonical(buf, x[k]); err != nil {
				return err
			}
		}
		buf.WriteByte('}')
		return nil
	case []any:
		buf.WriteByte('[') // R8: declaration order, never sorted
		for i, el := range x {
			if i > 0 {
				buf.WriteByte(',')
			}
			if err := writeCustodyCanonical(buf, el); err != nil {
				return err
			}
		}
		buf.WriteByte(']')
		return nil
	case string:
		return writeCustodyString(buf, x)
	case bool:
		if x {
			buf.WriteString("true")
		} else {
			buf.WriteString("false")
		}
		return nil
	case int:
		buf.WriteString(strconv.FormatInt(int64(x), 10))
		return nil
	case int32:
		buf.WriteString(strconv.FormatInt(int64(x), 10))
		return nil
	case int64:
		buf.WriteString(strconv.FormatInt(x, 10)) // R6: plain decimal
		return nil
	case float64:
		// Payloads are built natively with int64; floats appear only after a JSON decode
		// round-trip. Integral floats re-encode as R6 integers; anything else is rejected.
		if x == math.Trunc(x) && math.Abs(x) <= 1<<53 {
			buf.WriteString(strconv.FormatInt(int64(x), 10))
			return nil
		}
		return fmt.Errorf("custody: non-integral number %v has no canonical encoding (R6)", x)
	default:
		return fmt.Errorf("custody: type %T has no canonical encoding", v)
	}
}

// writeCustodyString emits a JSON string in UTF-8 without HTML escaping and without \uXXXX
// escapes for code points below U+0080 except the mandatory control/quote/backslash escapes (R5).
func writeCustodyString(buf *bytes.Buffer, s string) error {
	var tmp bytes.Buffer
	enc := json.NewEncoder(&tmp)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(s); err != nil {
		return fmt.Errorf("custody: string encode: %w", err)
	}
	buf.Write(bytes.TrimRight(tmp.Bytes(), "\n"))
	return nil
}

// CustodyEventHash computes lowercase-hex SHA-256 over the canonical payload with "eventHash"
// ALWAYS excluded (it is the output). previousHash, when the event type carries one, must
// already be present in fields — including the genesis empty string "".
func CustodyEventHash(fields map[string]any) (string, error) {
	canon := make(map[string]any, len(fields))
	for k, v := range fields {
		if k == "eventHash" {
			continue // unconditionally excluded from its own canonical payload
		}
		canon[k] = v
	}
	b, err := CustodyCanonicalJSON(canon)
	if err != nil {
		return "", err
	}
	sum := sha256.Sum256(b)
	return hex.EncodeToString(sum[:]), nil
}

// CustodyVerifyEvent recomputes the hash of a stored payload (with eventHash present) and
// reports whether it matches the recorded value.
func CustodyVerifyEvent(fields map[string]any) (bool, error) {
	recorded, _ := fields["eventHash"].(string)
	if recorded == "" {
		return false, fmt.Errorf("custody: event has no recorded eventHash")
	}
	got, err := CustodyEventHash(fields)
	if err != nil {
		return false, err
	}
	return got == recorded, nil
}

File: finalyearproject/repos/dkd-platform-libs/sdk/typescript/src/custody_hash.ts
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// CustodyHash Specification v2 — DM §2 (RFC-8785 subset R1-R9). One of five byte-identical
// runtime implementations; the shared gate is sdk/testvectors/custodyhash_vectors.json (PL-01).
import { createHash } from "node:crypto";

/** A JSON-ish value permitted in a custody event payload. Money/time arrive as bigint/number. */
export type CanonicalValue =
  | string
  | number
  | bigint
  | boolean
  | null
  | undefined
  | CanonicalValue[]
  | { [k: string]: CanonicalValue };

const CTRL_ESCAPES: Record<number, string> = {
  0x08: "\\b", 0x09: "\\t", 0x0a: "\\n", 0x0c: "\\f", 0x0d: "\\r",
};

const UTF8 = new TextEncoder();

/** R3 key ordering: ascending by UTF-8 byte value (not UTF-16 code unit). */
function byUtf8Bytes(a: string, b: string): number {
  const ba = UTF8.encode(a);
  const bb = UTF8.encode(b);
  const n = Math.min(ba.length, bb.length);
  for (let i = 0; i < n; i++) {
    if (ba[i] !== bb[i]) return ba[i] - bb[i];
  }
  return ba.length - bb.length;
}

/** Serialize `value` per CustodyHash Spec v2 rules R1-R9 (see DM §2). */
export function canonical(value: CanonicalValue): string {
  if (value === null || value === undefined) {
    throw new Error("custody: null forbidden outside omitted object members (R2)");
  }
  if (typeof value === "boolean") {
    return value ? "true" : "false"; // R7
  }
  if (typeof value === "bigint") {
    return value.toString(10); // R6 plain decimal
  }
  if (typeof value === "number") {
    // int64 values > 2^53 lose precision as JS `number` and MUST be passed as bigint.
    // The safe-integer boundary (<= 2^53, inclusive) matches the Go/Python oracle exactly.
    if (!Number.isInteger(value) || Math.abs(value) > 2 ** 53) {
      throw new Error(`custody: number ${value} has no canonical encoding — use bigint for int64 (R6)`);
    }
    return value.toString(10); // R6
  }
  if (typeof value === "string") {
    return encString(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(canonical).join(",") + "]"; // R8 order preserved, R9 recurse
  }
  if (typeof value === "object") {
    const keys = Object.keys(value)
      .filter((k) => value[k] !== null && value[k] !== undefined) // R2 omit null/absent
      .sort(byUtf8Bytes); // R3
    return "{" + keys.map((k) => encString(k) + ":" + canonical(value[k])).join(",") + "}"; // R4
  }
  throw new TypeError(`custody: type ${typeof value} has no canonical encoding`);
}

/** R5 — UTF-8, no HTML escaping, no \\uXXXX for code points >= U+0080; only mandatory escapes. */
function encString(s: string): string {
  let out = '"';
  for (const ch of s) {
    const o = ch.codePointAt(0)!;
    if (ch === '"') out += '\\"';
    else if (ch === "\\") out += "\\\\";
    else if (CTRL_ESCAPES[o] !== undefined) out += CTRL_ESCAPES[o];
    else if (o < 0x20) out += "\\u" + o.toString(16).padStart(4, "0");
    else out += ch; // literal UTF-8 (incl. <, >, &, Bangla, emoji)
  }
  return out + '"';
}

/** lowercase-hex SHA-256 over canonical(fields) with `eventHash` unconditionally excluded. */
export function eventHash(fields: Record<string, CanonicalValue>): string {
  const canon: Record<string, CanonicalValue> = {};
  for (const [k, v] of Object.entries(fields)) {
    if (k !== "eventHash") canon[k] = v;
  }
  return createHash("sha256").update(canonical(canon), "utf8").digest("hex");
}

/** Recompute the hash of a stored payload (with eventHash present) and report a match. */
export function verifyEvent(fields: Record<string, CanonicalValue>): boolean {
  const recorded = fields["eventHash"];
  if (typeof recorded !== "string" || recorded === "") {
    throw new Error("custody: event has no recorded eventHash");
  }
  return eventHash(fields) === recorded;
}
